src/library_egl.js

```diff
@@ -367,7 +367,7 @@ var LibraryEGL = {
 
     EGL.context = GL.createContext(Module['canvas'], EGL.contextAttributes);
 
-    if (EGL.context != 0) {
+    if (EGL.context > 0) {
```

Comment:
what was the bug here? is a null returned on failure, or something like that?

Reply:
(I ask because it seems a little surprising to me to see `> 0` here - what is the type of that value?)

Reply:
the bug is that EGL's process declared success even if the result of `GL.createContext `was negative, although a negative result is considered a failure. 
A valid OpenGL handle (for contexts and other objects) is strictly positive, and regardless of OpenGL, Emscripten's wrapping of OpenGL's handles uses the handles as indices into the arrays maintained by the `GL` singleton (see `getNewId` in library_webgl.js:230 for example. So trying to use a negative value would cause a crash, due to `undefined` being returned.

Reply:
The return value is an integer - an OpenGL handle.

Reply:
Where does a negative value come from, though? It seems like we should return 0 on failure from the function we call, shouldn't we? If so, then the bug might be in there, and the proper fix to avoid returning a negative number. But maybe I'm not seeing the whole picture yet.

Reply:
I read a bunch of code and still can't see where a negative number could come from..?

Reply:
See the change I added - `createContext` is called as part of `emscripten_webgl_create_context`, which according to the docs - 
`On success, a strictly positive value that represents a handle to the created context. On failure, a negative number that can be cast to an EMSCRIPTEN_RESULT field to get the reason why the context creation failed.`
0 is `EMSCRIPTEN_RESULT_SUCCESS`, which isn't the correct - the process failed.

If you'd prefer, `emscripten_webgl_create_context` can check if the value returned from `createContext` is 0, and then return `EMSCRIPTEN_RESULT_FAILED`, but IMO letting the internal functions return the correct error code for each failure reason allows for more precise errors.

Reply:
and in case, as I mentioned for the EGL code - negative numbers are unworkable as handles, and so should be treated as errors.

Reply:
Ok, I think I see. So `GL.createContext` used to always return 0 or a positive valid value. And this PR changes that, so we need to handle the new possible negative values in all places that call `GL.createContext`, is that right?

If so then it seems like a risky change. For example the call in `src/library_browser.js` looks wrong,
```javascript
contextHandle = GL.createContext(canvas, contextAttributes);
if (contextHandle) {
  ctx = GL.getContext(contextHandle).GLctx;
}
```
There may also be out of tree users of that API. Overall, I think it may not be worth changing this API's return value - is there a significant benefit that could justify that?

Reply:
This is a continuation of [this PR](https://github.com/emscripten-core/emscripten/pull/10982). The justification for both PRs is that ATM once you have issues with context creation, there's no way to understand what's the source of the issue and how to fix it - the lack of error codes means that the only way to debug these issues is to go into emscripten code in compiled in debug mode, and walkthrough the functions. 

I'd be happy to fix the call in `library_browser.js`, too. Again - every call that implicitly assumes that the return value of `createContext` is non-negative is dangerous, since negative values are unusable. IMO It's better to explicitly protect against negative return values.

Reply:
conceptually, since createContext doesn't have a well-documented API, I'd argue it should conform to `emscripten_webgl_do_create_context`, which returns handle or error code.

Reply:
Ah, I think I missed this important detail when reviewing that PR earlier, then. Sorry about that! Good point.

Thinking about this now, another option might be to use `GL_DEBUG` to debug these issues - it shows useful logging for each of them. So just building with that flag, running, and viewing the dev console should work - is that not good enough for some reason? (I'm trying to fully understand the need here.)

Reply:
I forgot about that flag, that's true, but assuming that you don't ship production code with `GL_DEBUG` turned on, you'd still have the issue of insufficient logs when this issue happens in production.

IMO The crux of the issue is that currently this will break code that until now only checked 0 as an error code, and didn't handle negative numbers, but that code was a bug waiting to happen, since negative numbers are unusable as handles.
I'm fine with hunting down and fixing all of these situations in emscripten's code, but if you view such breakage as no-go, I'll open a PR on `emscripten_webgl_do_create_context` that checks if `createContext` returns zero, and replaces the result with an informative error code. Would you prefer that?

Reply:
Good question - I'm not sure myself. Both paths have benefits, as you said.

cc @juj for feedback here.

Reply:
ping @juj

Reply:
(I believe @juj has been on vacation in recent weeks, but I think he'll be back soon.)

Reply:
Apologies for the delay, was away on summer vacation.

This change is a bit problematic. The core issue here is that the Emscripten WebGL context handle that is returned is a pointer obtained via a call to malloc:

https://github.com/emscripten-core/emscripten/blob/df5f6d150a4206b6935346b51fbd84da664c4b58/src/library_webgl.js#L941

This is so that in multithreaded applications the different threads can have a coherent access to the context handle for eglMakeCurrent()/emscripten_webgl_make_current() purposes.

Currently browsers still have the limitation that valid wasm heap pointer addresses are constrained to signed 32-bit integers (2GB range), but this is actively being worked on in the browsers. After that change, pointers will have proper unsigned 32-bit range.

When that happens, these functions can return values that are (seemingly) negative, but still correspond to valid WebGL context handles. So in that case, we will no longer be able to overload the return value to mean <0: specified error code, ==0: unspecified error, >0: valid context handle.

I think the best action here would be to restrict the current GL contextion creation functions to return 0 on failure, and non-zero on success; then update/fix up the documentation to reflect this. If desirable, then create a second WebGL context creation function

```c
extern EMSCRIPTEN_RESULT emscripten_webgl_create_context_ex(const char *target, const EmscriptenWebGLContextAttributes *attributes, EMSCRIPTEN_WEBGL_CONTEXT_HANDLE *outputHandle);
```

which will have space to return both the error code and the context output handle.

Then retain the functions GL.registerContext/GL.createContext/emscripten_webgl_create_context to not return GL context creation error codes, and have the function `emscripten_webgl_create_context_ex` do that instead.

That way this feature can be compatible with 4GB heap support in the future.

> A valid OpenGL handle (for contexts and other objects) is strictly positive

This is true for OpenGL handles, but a GL context is not an OpenGL handle; GL spec does not specify GL anything about GL context objects, but e.g. EGL does, and our own HTML5 WebGL "spec"/API does. So if we wanted, we could spec HTML5 WebGL API valid context handles to be numbers that are a multiple of 37 or something else arbitrary.

> Emscripten's wrapping of OpenGL's handles uses the handles as indices into the arrays maintained by the `GL` singleton (see `getNewId` in library_webgl.js:230 for example.

This is only true for singlethreaded builds, and there just for the reason of saving code space. In multithreaded builds, a GL context is a pointer obtained via malloc (like mentioned above).

> which according to the docs -
> `On success, a strictly positive value that represents a handle to the created context. On failure, a negative number that can be cast to an EMSCRIPTEN_RESULT field to get the reason why the context creation failed.`

It looks like we should definitely change the docs here to say

```
On success, a strictly positive value that represents a handle to the created context. On failure, zero is returned. In such case, build with GL_DEBUG to obtain more information about the GL context creation error, or use the function emscripten_webgl_create_context_ex to obtain an error code back to the application.
```

This means that we should revert the PR #10982, and reserve those changes to the new function `emscripten_webgl_create_context_ex`.

> If you'd prefer, `emscripten_webgl_create_context` can check if the value returned from `createContext` is 0, and then return `EMSCRIPTEN_RESULT_FAILED`, but IMO letting the internal functions return the correct error code for each failure reason allows for more precise errors.

For smallest code size, we can actually change `GL.createContext` to just `return;` (to return undefined) on failure, and then have `emscripten_webgl_create_context` test against falsyness, and `emscripten_webgl_create_context_ex` can then wrap to the correct error code.

This may seem really odd thing to do, though `emscripten_webgl_create_context`/`GL.createContext` is used on some 64K demoscene-like entries, where every generated byte counts, so we can help those usecases out.

> conceptually, since createContext doesn't have a well-documented API

Agreed. The intent here is that the function is internal, not to be called by any user code, hence the lack of careful documentation. For GL.createContext I think return undefined on failure, non-zero on success is a good way to go.

> IMO The crux of the issue is that currently this will break code that until now only checked 0 as an error code, and didn't handle negative numbers, but that code was a bug waiting to happen, since negative numbers are unusable as handles.

Once 4GB support is toggled, the 32-bit negative numbers are just an issue of calling code interpreting them as signed integers. They will be usable handles all the same.

Reply:
Fair enough. This seems like a pretty significant refactor - I'd like to help, where possible, but I don't think I can take this entirely by myself.

Reply:
Thanks @juj!

Sorry for the confusion @Shachlan , I didn't realize all this when I merged #10982. Yes, assuming pointers are positive signed 32-bit integers is a problem.

Sounds like the first step here is to revert that PR, can you please open a new PR for that?

Reply:
@kripken opened https://github.com/emscripten-core/emscripten/pull/11858, closing this PR.

I guess the next step is to update the docs and signature of `emscripten_webgl_create_context`, and maybe add the recommended `emscripten_webgl_create_context_ex`.